File: CokeLee777/src/com/leetcode/greedy/videostitching/VideoStitching.java
package com.leetcode.greedy.videostitching;

import java.util.Arrays;
import java.util.Comparator;

public class VideoStitching {

    static class MySourceCode{

        public int videoStitching(int[][] clips, int time) {
            // TODO : 시작시간을 기준으로 오름차순 정렬, 같으면 끝나는 시간이 큰 순서대로 정렬
            Arrays.sort(clips, (o1, o2) -> {
                if(o1[0] < o2[0]) return -1;
                else if(o1[0] == o2[0]) return o2[1] - o1[1];
                else return 1;
            });

            //시작부분이 존재하지 않는다면 -1반환
            if(clips[0][0] != 0) return -1;

            int segment = 1;
            int[][] editedVideo = new int[clips.length+1][2];
            for (int[] clip : clips) {
                // TODO : 다음 편집점이 필요하다면 -> 현재 클립이 전 클립과 이어지지 않고 끝나는 시간이 더 뒤에 끝난다면
                if(editedVideo[segment-1][1] < clip[0] && editedVideo[segment][1] < clip[1]) segment++;
                // TODO : 전의 편집점과 이어지고, 더 늦게 끝나는 편집점이라면 -> 세그먼트 업데이트
                if(clip[1] > editedVideo[segment][1]){
                    editedVideo[segment][0] = clip[0];
                    editedVideo[segment][1] = clip[1];
                }

                // TODO : 런타임 시간이 이어지지 않는다면 종료
                if(editedVideo[segment-1][1] < editedVideo[segment][0]) return -1;

                // TODO : 편집 중간에 이미 원하는 시간만큼의 비디오 클립을 완성했다면 종료
                if(editedVideo[segment][1] >= time) return segment;
            }

            // TODO : 비디오 클립을 완성할 수 없다면 -1반환
            return editedVideo[segment][1] >= time ? segment : -1;
        }
    }
}

File: dongdaang/greedy/video_stitching/Video_Stitching.py
# https://leetcode.com/problems/video-stitching/


class Solution:
    def videoStitching(self, clips, time):
        clips.sort()
        
        if clips[0][0] != 0:
            return -1
        
        N = len(clips)
        used = [False] * N      #사용된 clip 표시를 위한 리스트
        covered = clips[0]      #현재까지 커버된 video 파트
        used[0] = True
        res = 0
        
        for i in range(1, N):
            if covered[1] >= time:
                break
            if clips[i][1] > covered[1] and clips[i][0] <= covered[1]:
                covered = [0, clips[i][1]]
                used[i] = True
                for j in range(i):
                    if used[j]:
                        if clips[j][0] == clips[i][0]:
                            for k in range(j, i):
                                used[k] = False
                        elif clips[j][0] < clips[i][0] <= clips[j][1]:
                            for k in range(j + 1, i):
                                used[k] = False
        
        if covered[1] < time:
            return -1
        
        res = used.count(True)
        
        return res

a = Solution()
print(a.videoStitching([[5,7],[1,8],[0,0],[2,3],[4,5],[0,6],[5,10],[7,10]], 5))


##############솔루션################

# def videoStitching(self, clips, T):
#     end, end2, res = -1, 0, 0
#     for i, j in sorted(clips):
#         if end2 >= T or i > end2:
#             break
#         elif end < i <= end2:
#             res, end = res + 1, end2
#         end2 = max(end2, j)
#     return res if end2 >= T else -1